### PR TITLE
[Tooling] Release-on-CI automation improvements

### PR DESCRIPTION
**Pull request by iangmaia**
This PR implements a few additions for our release automation on CI:

- Automatically move PRs from the current release milestone to the next one
- Uses the shared `create_release_backmerge_pull_request` Action from [release-toolkit](https://github.com/wordpress-mobile/release-toolkit), assigning a milestone to the backmerge PR.
- Make sure release tasks won't fail due to milestones already being closed
- Annotate builds triggering other builds with their URL
- Close hotfix milestone
- Updated release pipelines to not allow manual retry
- Added `ensure_branch_does_not_exist` in the beginning of `code_freeze`, to avoid unexpected behavior and errors.
- ~Removed the `:trigger_` testing lanes, as testing can happen either via Rv2 or the Buildkite web UI.~ (moved to the PR https://github.com/wordpress-mobile/WordPress-iOS/pull/23589)

**Update**:
- Merged https://github.com/wordpress-mobile/WordPress-iOS/pull/23592 into this PR (without realizing it wasn't for `trunk`), so it now also contains release publishing.

### CODE REVIEW

File: fastlane/lanes/release.rb [hunk 1]
```diff
@@ -354,19 +356,28 @@
...
     ensure_git_status_clean
 
-    # Pull the latest hotfix release branch changes
-    git_pull
+    hotfix_version = release_version_current
 
-    UI.important("Triggering hotfix build for version: #{release_version_current}")
-    UI.user_error!('Aborted by user request') unless options[:skip_confirm] || UI.confirm('Do you want to continue?')
+    UI.important("Triggering hotfix build for version: #{hotfix_version}")
+    UI.user_error!("Terminating as requested. Don't forget to run the remainder of this automation manually.") unless skip_confirm || UI.confirm('Do you want to continue?')
```
Comment (dangermattic) [inline fastlane/lanes/release.rb:366]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/3ea5ce7fb46853fc78e2b0ca2196de39e75cd487/fastlane/lanes/release.rb#L366"></span>Style/IfUnlessModifier: Modifier form of <code>unless</code> makes the line too long.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/release.rb [hunk 2]
```diff
@@ -379,53 +390,45 @@
...
-    lint_localizations(allow_retry: skip_user_confirmation == false)
+    lint_localizations(allow_retry: skip_confirm == false)
 
     bump_build_codes
 
-    unless skip_user_confirmation || UI.confirm('Ready to push changes to remote and trigger the release build?')
-      UI.message("Terminating as requested. Don't forget to run the remainder of this automation manually.")
-      next
-    end
+    UI.user_error!("Terminating as requested. Don't forget to run the remainder of this automation manually.") unless skip_confirm || UI.confirm('Ready to push changes to remote and trigger the release build?')
```
Comment (dangermattic) [inline fastlane/lanes/release.rb:409]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/3ea5ce7fb46853fc78e2b0ca2196de39e75cd487/fastlane/lanes/release.rb#L409"></span>Style/IfUnlessModifier: Modifier form of <code>unless</code> makes the line too long.</td>
    </tr>
  </tbody>
</table>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/296">25.4</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: .buildkite/release-builds.yml [hunk 3]
```diff
@@ -11,10 +11,18 @@ steps:
     command: ".buildkite/commands/release-build-wordpress.sh $BETA_RELEASE"
     plugins: [$CI_TOOLKIT_PLUGIN]
     notify:
-    - slack: "#build-and-ship"
+      - slack: "#build-and-ship"
```
Comment (mokagio) [inline .buildkite/release-builds.yml:14]: On seeing this I wondered "How has it worked all this time?!" but it turns out YAML (or at least the Ruby parser and Buildkite's parser) are flexible on array indentation:

image

Comment (mokagio) [inline .buildkite/release-builds.yml:14]: On seeing this I wondered "How has it worked all this time?!" but it turns out YAML (or at least the Ruby parser and Buildkite's parser) are flexible on array indentation:

image

File: .buildkite/release-builds.yml [hunk 4]
```diff
@@ -11,10 +11,18 @@ steps:
...
+    retry:
+      manual:
+        # If those jobs fail, one should always prefer re-triggering a new build from ReleaseV2 rather than retrying the individual job from Buildkite
+        allowed: false
 
   - label: ":jetpack: :testflight: Jetpack Release Build (App Store Connect)"
     command: ".buildkite/commands/release-build-jetpack.sh"
     plugins: [$CI_TOOLKIT_PLUGIN]
     notify:
     - slack: "#build-and-ship"
```
Comment (mokagio) [inline .buildkite/release-builds.yml]: For consistency:

```suggestion
    notify:
      - slack: "#build-and-ship"
```

File: fastlane/lanes/release.rb [hunk 5]
```diff
@@ -104,30 +107,44 @@
...
+      report_milestone_error(error_title: "Error freezing milestone `#{new_version}`: #{e.message}")
     end
 
-    set_milestone_frozen_marker(repository: GITHUB_REPO, milestone: new_version)
+    UI.message("Moved the following PRs to milestone #{release_version_next}: #{moved_prs.join(', ')}")
 
-    check_pods_references
+    # Annotate the build with the moved PRs
+    moved_prs_info = if moved_prs.empty?
+                       "👍 No open PR were targeting `#{new_version}` at the time of code-freeze"
```
Comment (mokagio) [inline fastlane/lanes/release.rb]: Nitpick: were requires a plural

```suggestion
                       "👍 No open PRs were targeting `#{new_version}` at the time of code-freeze"
```

File: fastlane/lanes/release.rb [hunk 6]
```diff
@@ -457,14 +506,17 @@
...
+  build_url = buildkite_trigger_build(
     buildkite_organization: BUILDKITE_ORGANIZATION,
     buildkite_pipeline: BUILDKITE_PIPELINE,
     branch: branch,
     environment: { BETA_RELEASE: beta },
     pipeline_file: 'release-builds.yml',
     message: beta ? 'Beta Builds' : 'Release Builds'
   )
+
+  message = "This build triggered a build on <code>#{branch}</code>:<br>- #{build_url}"
```
Comment (mokagio) [inline fastlane/lanes/release.rb]: Nitpick: Other messages us Markdown formatting.

```suggestion
  message = "This build triggered a build on `#{branch}`:\n\n- #{build_url}"
```

File: fastlane/lanes/release.rb [hunk 7]
```diff
@@ -379,53 +391,90 @@
...
-      title: "Merge #{version} release finalization"
-    )
-
+    pr_url = create_backmerge_pr
     message = <<~MESSAGE
       Release successfully finalized. Next, review and merge the [integration PR](#{pr_url}).
     MESSAGE
     buildkite_annotate(context: 'finalization-completed', style: 'success', message: message) if is_ci
+
+    remove_branch_protection(repository: GITHUB_REPO, branch: release_branch_name)
```
Comment (mokagio) [inline fastlane/lanes/release.rb]: ~~Should the step to remove branch protection be moved to the `publish_release` lane?~~

Update: I see the method to delete the branch below removes the branch protection, so I'd say we can remove this altogether

```suggestion
```

File: fastlane/lanes/release.rb [hunk 8]
```diff
@@ -457,14 +506,17 @@
...
     buildkite_organization: BUILDKITE_ORGANIZATION,
     buildkite_pipeline: BUILDKITE_PIPELINE,
     branch: branch,
     environment: { BETA_RELEASE: beta },
     pipeline_file: 'release-builds.yml',
     message: beta ? 'Beta Builds' : 'Release Builds'
   )
+
+  message = "This build triggered a build on <code>#{branch}</code>:<br>- #{build_url}"
+  buildkite_annotate(style: 'info', context: 'trigger-release-build', message: message) if is_ci
```
Comment (mokagio) [inline fastlane/lanes/release.rb]: A different nitpick. Given we don't use `message` unless we are in CI:

```suggestion
  return unless is_ci
  
  message = "This build triggered a build on `#{branch}`:\n\n- #{build_url}"
  buildkite_annotate(style: 'info', context: 'trigger-release-build', message: message)
```

**Review by mokagio — APPROVED** [review]
Looking good. However it's always hard to tell withe these, short of running them 😞 Easy to miss a rename or trivial things like that.

None of my suggestions and nitpicks are blocking, so leaving it up to you whether to merge or to apply first.

Probably a conversation for an RFC in P2, but I wonder what you think of my approach [from Simplenote Android](https://github.com/Automattic/simplenote-android/commit/ec7a563ec76d850cf7ea276740b82fb0cadb0921) where:

- `create_backmerge_pr` is `create_backmerge_pr!` to account for the `UI.user_error!` call
- We also have `create_backmerge_prs!` to differentiate between lanes that expect one backmerge PR only vs lanes that expect more (e.g. finalizing a hotfix)

I think it would be quite unlikely, especially in WordPress which now releases slowly, but if by chance we get a `pr_url` array instead of a single element, the annotation we then print it in might be confusing.

File: .buildkite/release-builds.yml (hunk 3 above)
Comment (iangmaia) [reply .buildkite/release-builds.yml:14]: Yep, I was also surprised 🙂

File: fastlane/lanes/release.rb (hunk 7 above)
Comment (iangmaia) [reply fastlane/lanes/release.rb]: Ah, nice catch! I believe this is a leftover from when apps didn't have a release_publish lane.

File: fastlane/lanes/release.rb [hunk 9]
```diff
@@ -457,14 +504,19 @@
...
     buildkite_organization: BUILDKITE_ORGANIZATION,
     buildkite_pipeline: BUILDKITE_PIPELINE,
     branch: branch,
     environment: { BETA_RELEASE: beta },
     pipeline_file: 'release-builds.yml',
     message: beta ? 'Beta Builds' : 'Release Builds'
   )
+
+  return unless is_ci
+
```
Comment (dangermattic) [inline fastlane/lanes/release.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-iOS/blob/b318b700f75303b75dbea0608be7c6cf0f260036/fastlane/lanes/release.rb#L517"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/release.rb (hunk 6 above)
Comment (iangmaia) [reply fastlane/lanes/release.rb]: I've updated the suggestion https://github.com/wordpress-mobile/WordPress-iOS/pull/23585#discussion_r1766222155 to bundle this change together with the early return.

File: .buildkite/release-builds.yml (hunk 3 above)
Comment (AliSoftware) [reply .buildkite/release-builds.yml:14]: Was surprised as well 😅 

💡  Tip of the day: If you want to test how `buildkite-agent` itself would interpret your YAML (and thus be sure to rely on its implementation of YAML parser), you can use `buildkite-agent pipeline upload --dry-run` to print its interpretation of the pipeline. It prints the interpreted pipeline to stdout—in JSON by default, but you can use `--format=yaml` if you prefer YAML output that would be closer to the input file you provided.

For example, using a pipe to pass the YML to parse:
```bash
# Note: you need to provide the `-i` flag (but not add the `-t` flag!) for `docker run` to be able to read from STDIN through the pipe properly
echo "notify:\n- x" | docker run --rm -i buildkite/agent pipeline upload --dry-run
cat .buildkite/release-builds.yml | docker run --rm -i buildkite/agent pipeline upload --dry-run --format=yaml
```
Or if you want to pass it a file path to parse:
```bash
docker run --rm -v ./:/app -w /app buildkite/agent pipeline upload --dry-run --format=yaml .buildkite/pipeline.yml
```

File: fastlane/lanes/release.rb (hunk 9 above)
Comment (mokagio) [reply fastlane/lanes/release.rb]: 🤔 where is the trailing whitespace RuboCop / Danger detects?

image

Comment (mokagio) [reply fastlane/lanes/release.rb]: It was on line 517, even though the GitHub diff didn't show it

image

536b1c0

File: fastlane/lanes/localization.rb [hunk 10]
```diff
@@ -243,11 +243,7 @@ def generate_strings_file(gutenberg_path:, derived_data_path:)
 
     push_to_git_remote(tags: false)
 
-    pr_url = create_release_management_pull_request(
-      release_version: release_version,
-      base_branch: compute_release_branch_name(options: options, version: release_version),
-      title: "Merge editorialized release notes in #{release_version}"
-    )
+    pr_url = create_backmerge_pr
```
Comment (oguzkocer) [inline fastlane/lanes/localization.rb:246]: @iangmaia @mokagio Unfortunately this change is not correct. Please see [this comment from WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/21229/files#r1800099465) for the same issue.

Comment (iangmaia) [reply fastlane/lanes/localization.rb:246]: For the record, this PR should fix this issue: https://github.com/wordpress-mobile/WordPress-iOS/pull/23758

**Pull request merged**